### DISCUSSION TITLE
[Bug]: Self-hosted runners at the Enterprise level fail to detect queued jobs

### DISCUSSION THREAD

**Discussion by kathleenfrench** [community/community#27119] [Enterprise]
<p>Hi! I believe I’ve discovered a bug in how the actions runner is currently handling routing/queuing active workflows to available runners.</p>
<p>I opened an issue in the ‘actions/runner’ repository here (<a href="https://github.com/actions/runner/issues/1059" class="inline-onebox" rel="noopener nofollow ugc">Runner.Listener failing to detect queued workflows/jobs for Enterprise runners · Issue #1059 · actions/runner · GitHub</a>) and was encouraged to follow up in this forum instead.</p>
<p>There is a lot more additional context in the issue I originally posted here (<a href="https://github.com/summerwind/actions-runner-controller/issues/470" class="inline-onebox" rel="noopener nofollow ugc">enterprise autoscaling issues, indefinitely queued jobs within workflows · Issue #470 · summerwind/actions-runner-controller · GitHub</a>) on the subject as well.</p>
<p>To boil it down, though, the disconnect seems to be stemming from a failure in the job message queue to retry allocating queued workflows to newly available, and otherwise compatible, runners. This becomes in issue when attempting to autoscale actions as a service.</p>
<p><strong>Expected Behavior</strong></p>
<ul>
<li>1 self-hosted enterprise worker is up</li>
<li>1 workflow is triggered containing 4 jobs</li>
<li>The 1 available runner accepts 1 of the jobs, 3 jobs are now queued</li>
<li>A scale-up is triggered for the enterprise runners - now 4 runners are available (this autoscaling is currently working via the custom kubernetes controller we are using from <a href="https://github.com/summerwind/actions-runner-controller" rel="noopener nofollow ugc">summerwind/actions-runner-controller</a>)</li>
<li>After the scale-up, the newly available runners detect there are 3 queued jobs and run them</li>
</ul>
<p><strong>Actual Behavior</strong></p>
<ul>
<li>1 self-hosted enterprise worker is up</li>
<li>1 workflow is triggered containing 4 jobs</li>
<li>The 1 available runner accepts 1 of the jobs, 3 jobs are now queued</li>
<li>Scale-up brings the available runners to 4 - Github API accurately reflects that the 3 additional runners are online and not busy</li>
<li>The 3 queued jobs never run and indefinitely hang in a sort of queue purgatory, despite the new runners indicating in the logs that they are listening for jobs. If a new workflow is triggered, these runners detect those jobs without issue, but the queued jobs are never picked up.</li>
</ul>
<p>This is the case whether it’s jobs of the same workflow, or if a separate workflow entirely is triggered and not picked up, so it doesn’t appear to be a bug specific to jobs themselves, rather it’s any sort of task.</p>
<p>The self-hosted runner documentation states that the order of operations is to first search for available runners at the repository level, then the organization level, then for the first available runner enterprise-wide that satisfies the constraints set in the labels/groups/os, for that particular workflow. This does not appear to be the case for enterprise-level runners alone as they will only detect jobs that are immediately assigned, i.e. if in the example above I had 4 enterprise runners up it would have run without a problem.</p>
<p>So, I tried adding in an ‘organization’-scoped runner <em>in addition</em> to the enterprise runner. In this case there <em>is</em> a difference in behavior - the organization-scoped runners scale-up (as the enterprise runners do), but unlike the enterprise runners the organization-scoped runners <em>do</em> detect jobs that are actively waiting in the queue.</p>
<p><strong>There is an important caveat to this, though, as it assumes at least one of those jobs was captured by an available org-level runner at the time the workflow was triggered</strong>. If, for instance, none are available at the time the workflow is triggered, similar to the enterprise failure to detect queued jobs, those jobs are never detected by the org-runners.</p>
<p>Again, for all of the above instances the Github API accurately reflects that these self-hosted runners are online and available. There are no errors in the runner logs - <code>stdout</code> or <code>/runner/_diag/</code> - that speak to the issue - everything reflects the expected state there as well.</p>
<p>The only way to handle these queued jobs is to cancel the entire workflow manually.</p>
<p>Please let me know if I can provide any more information on my end!</p>

**Comment by hross** [comment, marked as answer]
<p>:wave: Software manager for the team that owns this area of the product…</p>
<ol>
<li>Thanks for such a detailed write up of the issue. Much appreciated.</li>
<li>I am following the thread over here: <a href="https://github.com/summerwind/actions-runner-controller/issues/470" class="inline-onebox">enterprise autoscaling issues, indefinitely queued jobs within workflows · Issue #470 · summerwind/actions-runner-controller · GitHub</a> on the controller as well.</li>
</ol>
<p>Yes, this is a bug in the assignment logic for the service (or at least, undocumented behavior that doesn’t do what you’d expect). We are planning to ship a fix for this behavior in the next few weeks. Given that it is job assignment logic, though, it might take us some time to test and verify the behavior as we roll it out (breaking job assignment would be pretty bad :smile:.</p>
<p>Also I appreciate that you followed up in the forums even though you had to write things up twice. We are trying to clean up the runner repo and raise things in the appropriate places if we can.</p>
<p>If you have any further questions in this area please feel free to reach out directly to me via email if you like (handle at github).</p>
<p>We are investing in making queuing/assignment/runs better this quarter so I would love feedback.</p>
<p>Thanks!</p>

**Reply by pacha5065** [reply]
Do we have any fix available for this issue? It seems we are experiencing a similar issue in our enterprise now.

**Comment by kathleenfrench** [comment]
<p>Great! Sent you an e-mail - I’ll close this out in the meantime as there are plans to resolve these issues in the near-future. :+1:</p>
